554: keyboard shortcuts

toolstack opened the pull request:
Resolves #554.

toolstack committed:
Add new keyboard shortcuts for the editor.

Ctrl Up = Previous editor
Ctrl Down = Next editor
Ctrl Enter = Copy original
Ctrl - = Reject
Ctrl + = Approve

toolstack committed:
Declare variables.

toolstack committed:
Coding standards cleanup.

toolstack committed:
Update minified editor.js.

toolstack committed:
Coding standards cleanup.

akirk commented on assets/js/editor.js:
```diff
@@ -129,6 +129,28 @@ $gp.editor = (
+					original = $( '.editor:visible' ).find( '.original' );
+
+					target.val( original.text() );
+				} else if ( 107 === e.keyCode && e.ctrlKey ) { // Ctrl-+ = Approve.
```
`Ctrl-+` zooms in the webpage in most browsers in Windows, see [Windows > Webpage shortcuts > Make everything on the page bigger](https://support.google.com/chrome/answer/157179)

akirk replied:
What about `Cmd+Shift+a` resp. `Ctrl+Shift+a`?

toolstack replied:
I think we'll never avoid all shortcuts used by every os or user, we already hijack PgUp/PgDn.  This is also only active within the editor box, so I don't see it as a big deal.

akirk commented on assets/js/editor.js:
```diff
@@ -129,6 +129,28 @@ $gp.editor = (
+					if ( approve.length > 0 ) {
+						approve.trigger( 'click' );
+					}
+				} else if ( 109 === e.keyCode && e.ctrlKey ) { // Ctrl-- = Reject.
```
Same as above, `Ctrl--` zooms out the webpage in most browsers in Windows.

akirk replied:
What about `Cmd+Shift+r` resp. `Ctrl+Shift+r`?

toolstack replied:
See above.

ocean90 commented on assets/js/editor.js:
```diff
@@ -103,15 +103,15 @@ $gp.editor = (
+				if ( 27 === e.keyCode ) { // Escape = Next editor.
 					$gp.editor.hide();
-				} else if ( 33 === e.keyCode ) {
+				} else if ( 33 === e.keyCode || ( 38 === e.keyCode && e.ctrlKey ) ) { // Page Down or Ctrl-Up Arrow = Previous editor.
```
CTRL-[UP/DOWN/LEFT/RIGHT] are used by OS X to open Mission Control and to change spaces.

toolstack replied:
See above.

GaryJones commented:
Thanks for this @toolstack.

I'm not overly concerned with the keyboard shortcuts for Approve and Reject right now, since they are often done with the bulk checkboxes. It's the "Copy from original" that's useful for #554, though needs to take into account plural fields too.

Side note - GitHub Accepts Ctrl-Enter as Submit Comment. Maybe that's safe for GlotPress submission (instead of Shift-Enter), and use the Shift-Enter for Copy from original?

toolstack committed:
Make sure to only use the last original found.

This matches the current "Copy from original" behaviour.

toolstack committed:
Fix copy from original to select the correct plural.

Updated both keyboard shortcut and "Copy from original" link.

toolstack committed:
Add missing radix and declare variable.

toolstack committed:
Coding standards cleanup.

toolstack committed:
Update minified version.

toolstack committed:
Coding standards update.

toolstack commented:
@GaryJones I wouldn't want to change an existing shortcut and replace it with something else, that would be confusing for the users.

toolstack committed:
Add some compatibility with GlotDict shortcuts.

Specifically:

Ctrl-Shift-Z to cancel
Ctrl-Shift-B to copy original
Ctrl-Shift-A to approve
Ctrl-Shift-R to reject

ocean90 commented on gp-templates/helper-functions.php:
```diff
@@ -117,7 +117,7 @@ function textareas( $entry, $permissions, $index = 0 ) {
 			}
 			else {
-				printf( __( 'You <a href="%s">have to log in</a> to edit this translation.', 'glotpress' ), esc_url( wp_login_url( gp_url_current() ) ) );
+				printf( __( 'You <a href="%s">have to log in</a> to edit this translation.', 'glotpress' ), esc_url( wp_login_url() ) );
```
This reverts a previous commit.

toolstack replied:
Not sure how that slipped in there... will undo it.

ocean90 commented on assets/js/editor.js:
```diff
@@ -129,6 +129,31 @@ $gp.editor = (
 					} else {
 						$gp.editor.save( target.parents( 'tr.editor' ).find( 'button.ok' ) );
 					}
+				} else if ( ( 13 === e.keyCode && e.ctrlKey ) || ( 66 === e.keyCode && e.shiftKey && e.ctrlKey ) ) { // Ctrl-Enter or Ctrl-Shift-B = Copy original.
```
Shouldn't this just trigger a click event for `a.copy` so that the `copy()` method gets used?

toolstack replied:
It could, I'll update the code.

toolstack committed:
Undo incorrect change.

toolstack committed:
Replace dedicated logic with artificial click of copy original link.

toolstack committed:
Coding standards cleanup.

ocean90 commented on assets/js/editor.js:
```diff
@@ -129,6 +129,28 @@ $gp.editor = (
+						approve.trigger( 'click' );
+					}
+				} else if ( ( 109 === e.keyCode && e.ctrlKey ) || ( 82 === e.keyCode && e.shiftKey && e.ctrlKey ) ) { // Ctrl-- or Ctrl-Shift-R = Reject.
+					target = $( e.target );
```
`target` is unused.

ocean90 commented on assets/js/editor.js:
```diff
@@ -129,6 +129,28 @@ $gp.editor = (
+						copy.trigger( 'click' );
+					}
+				} else if ( ( 107 === e.keyCode && e.ctrlKey ) || ( 65 === e.keyCode && e.shiftKey && e.ctrlKey ) ) { // Ctrl-+ or Ctrl-Shift-A = Approve.
+					target = $( e.target );
```
`target` is unused.

toolstack committed:
Removed unused code.

GaryJones commented:
This is awesome - thank you so much for merging this - it's going to save `en_*` translators about half the time I reckon!
